Dmr logic experimental

dl1hrc opened the pull request:
Please take a look into callback function. I need a buffer to store the incoming data from the 1st call up until all data has been received. It look like that the t_buffer is reinitialized with the new data from buffer when the function is called up twice.

porst17 commented on src/async/audio/AsyncAudioCodecAmbe.cpp:
```diff
@@ -110,37 +137,174 @@ namespace {
+                                       (sigc, charactersReceived method) is only a part of the datastream 
+                                       so we must concat them when all data defined in stored_bufferlen 
+                                       has been reeived */
+              t_buffer = Buffer<>(buffer.data, buffer.length); /* try to store the partially received
```
`buffer.data` is an array allocated on the stack in [`SerialDevice::onIncomingData`](https://github.com/sm0svx/svxlink/blob/master/src/async/core/AsyncSerialDevice.cpp#L300-L315). Variables on the stack are cleaned up immediately when the scope they have been created in is left. This means that the data you store in `t_buffer` is overwritten on the next call by something different. The solution is do something like
```
t_buffer = Buffer<>(new char[reasonable_size],reasonable_size);
memcpy(t_buffer.data, buffer.data, buffer.length);
```
If the maximum size of `t_buffer` is known in advance, say `T_BUFFER_SIZE`, you can skip the creation of the new array and just initialize it once below. This also has the advantage of automatic cleanup when the `AudioCodecAmbeDv3k` is destructed and avoids annoying bookkeeping.

porst17 commented on src/async/audio/AsyncAudioCodecAmbe.cpp:
```diff
@@ -110,37 +137,174 @@ namespace {
-                assert(!"invalid conversion from float * to char *");
-                AudioDecoder::sinkWriteSamples((float *)unpacked.data, unpacked.length);
+              stored_bufferlen = 0;
+              t_buffer = Buffer<>(buffer.data, buffer.length);
```
Same here.

porst17 commented on src/async/audio/AsyncAudioCodecAmbe.cpp:
```diff
@@ -168,10 +333,16 @@ namespace {
+      STATE m_state;
+      uint32_t stored_bufferlen;
+      int act_framelen;
+      Buffer<> t_buffer;
```
If the maximum size of `t_buffer` is known in advance, you could initialize it here and save the periodic reinitialization above.

porst17 commented on src/async/audio/AsyncAudioCodecAmbe.cpp:
```diff
@@ -110,37 +137,174 @@ namespace {
+          Buffer<> dv3k_buffer = buffer;
+          
+          // check if the data buffer contain valid data
+          if (buffer.length < 4)
```
You should not raise an error here. If the length is too short, just concat to `t_buffer` until enough bytes for further processing are there. If `buffer.length == 0`, you may print a warning and return or just ignore the call and return. If you have at least 4 bytes in `t_buffer` and they are garbage, then you may raise an error and reinit or whatever is necessary.

porst17 reviewed (commented):
In the `callback` method, you should always check if you have sufficient data for further processing and otherwise copy the incoming data into a temporary array regardless of the number of incoming bytes (be it 2 or 2000000). If there is sufficient data, continue processing, otherwise return and wait for more data. Always copying the incoming data is not the most efficient approach, but my advice is to do so for the moment until the whole implementation is functional. Once everything works, you can start to optimize performance.

BTW: Is the maximum size of a Dv3k frame is known? This would make quite a few things easier.

Dloranger committed:
Adding new Config options related to enabling IDs

adding several config options to allow user to configure VOICE/CW/Custom
announcements during the short and long ID processes.  This allows the
settings to be on a channel by channel basis rather than having a single
file for all channels

sm0svx committed:
Add the COMBINE squelch type

There now is a new squelch type that can combine multiple other squelch
types using a logical expression.

Implementing this new feature required rewriting quite a lot of the
squelch infrastructure code. Watch out for changed behavior in the
squelch functionality.

sm0svx committed:
Devcal manual page fixes

sm0svx committed:
Update copyright messages

sm0svx committed:
Fix bash variable spelling error in svxlink_gpio_up

Also put the generated scripts in the build directory bin-dir and make
them executable.

sm0svx committed:
Make systemd files more configurable from CMake

- More systemd configuration variable paths are now settable from CMake
  variables
- Now possible to set LOGFILE and PIDFILE from the systemd unit
  configuration files
- No need to start applications under a shell

sm0svx committed:
Merge pull request #408 from phl0/fixTypo

Fix typo

sm0svx committed:
HIDRAW squelch detector bugfix

The squelch detection logic for HIDRAW squelch was wrong so it did not
work for active low devices.

This is a simpler fix for a couple of GitHub PRs.
Closes #402 and #452.

sm0svx committed:
Improved implementation for Async::AudioGenerator

sm0svx committed:
Use Async::AudioGenerator in AsyncAudioIO_demo

sm0svx committed:
Implement triangular wave in Async::AudioGenerator

sm0svx committed:
The EventHandler is now not dependent on Logic

sm0svx committed:
PR code cleanup

- Now whole path have to be given for announcement files
- Rename variables LONG_ANNOUNCE and SHORT_ANNOUNCE
  LONG_ANNOUNCE -> LONG_ANNOUNCE_FILE
  SHORT_ANNOUNCE -> SHORT_ANNOUNCE_FILE
- Remove CustomIdent audio clip context
  We don't need a CustomIdent context since the SHORT_ANNOUNCE_FILE and
  LONG_ANNOUNCE_FILE variables can be used to specify the full path to a
  file.
- Remove debug printouts
- Update ChangeLog and versions

dl1hrc committed:
windshear now with runway, error handling for metar txt file, fix some typo

sm0svx committed:
Bugfix: The EchoLink module could crash when encountering TCL errors

Solves issue #455.

dl1hrc committed:
Merge remote-tracking branch 'sm0svx/master' into metarbugfix

sm0svx committed:
Update ChangeLog and ModuleMetarInfo version

sm0svx committed:
Merge branch 'dl1hrc-metarbugfix'

sm0svx committed:
Merge branch 'Dloranger-Add-config-options-to-IDs'

sm0svx committed:
Merge branch 'maint' into 'master'

sm0svx committed:
Handle "make package" when not in git working copy

sm0svx committed:
New environment variable ASYNC_AUDIO_ALSA_ZEROFILL

AudioDeviceAlsa can now write zeros to the audio device on underflow. That
is, when the audio device is open but there is no audio to write zeros will
be written instead. Enable this behavior by setting the environment variable
ASYNC_AUDIO_ALSA_ZEROFILL=1.

sm0svx committed:
Optionally keep audio device open

Now possible to configure SvxLink to keep audio devices open once opened.
The normal behavior in SvxLink is to open and close audio devices as needed.
In a local RX and/or TX section set the variable AUDIO_DEV_KEEP_OPEN=1.

sm0svx committed:
Now possible to keep audio device open from start

It is now possible to configure SvxLink to keep an audio device open from
application start to exit. The previous fix did not open the audio device
until it was used to first time.

sm0svx committed:
Initial version of TG reflector support

YouriNL committed:
Update to required packages.

tcl8.6-dev -> tcl8.6-dev
libgcrypt11-dev -> libgcrypt20-dev

sm0svx committed:
Add stop timeout for systemd units

Add stop timeout for systemd units after which svxlink will be killed if not
exiting within 10 seconds.

sm0svx committed:
Fix typo in error printout EVDEV_NAME -> EVDEV_DEVNAME

sm0svx committed:
Merge branch 'master' of github.com:sm0svx/svxlink

sm0svx committed:
Add package dependency to Docker debian-build

Add new dependency to Docker debian-build for libcurl4-openssl-dev.

dl1hrc committed:
first step to merge with master

sm0svx committed:
ASYNC_AUDIO_ALSA_ZEROFILL added for systemd

Added the ASYNC_AUDIO_ALSA_ZEROFILL environment variable to the systemd
environment files for svxlink and remotetrx.

sm0svx committed:
Add dependency for libcurl-devel to Docker centos6-build

sm0svx committed:
Add dependency for libcurl-devel to Docker centos7-build

sm0svx committed:
Add dependency for libcurl-devel to Docker fedora-build

sm0svx committed:
Fix build problems in Docker ubuntu-build

- Set DEBIAN_FRONTEND=noninteractive to stop apt from asking questions
  during container build
- Add package dependency librtlsdr-dev

sm0svx committed:
Add missing header files

sm0svx committed:
Fix warning for newer compilers

sm0svx committed:
Replace memset with better C++ correct initialization

sm0svx committed:
Configurable SMTP server in ModuleTclVoiceMail

Now possible to specify which mail server to use in the voice mail module

sm0svx committed:
ModuleDtmfRepeater bugfix and reimplementation

Logic and timings were a bit strange and buggy in the DTMF repeater
module so the implementation was changed a bit.

sm0svx committed:
Fix memory handling problem in ModuleMetarInfo

Double free of regexp objects removed

sm0svx committed:
Fix memory leak in ModuleMetarInfo

The Http object was never freed.

sm0svx committed:
Set uninitialized variables in ModuleMetarInfo

Some struct members in a struct tm was uninitialized. It probably did
not cause any problems but made valgrind warn about access to
uninitialized memory.

sm0svx committed:
Add more valgrind suppressions

sm0svx committed:
Fix memory handling problem in ModuleMetarInfo

- Fix potential problem if there is a wathch on both read and write for
  a file descriptor. With the old implementation only one of the file
  descriptor watches would be stored and so the other one is lost.
- This also fix a problem with memory deallocation (leak) in a conneted
  signal for the file descriptor watch.

sm0svx committed:
Fix memory leak in ModuleMetarInfo

The url_queue was not freed on Http object destruction.

sm0svx committed:
ModuleMetarInfo: Correction for previous fix

The file descriptor watches never were enabled. This also concealed one
of the other problems where memory was not freed for the
FdWatch::activity signal. This problem is also fixed now.

sm0svx committed:
Better support for V1 clients

sm0svx committed:
Set TG via DTMF

sm0svx committed:
Fix compilation error

sm0svx committed:
Merge branch 'master' into reflector_tg

sm0svx committed:
Fix second talker lockout bug

sm0svx committed:
Handle TG switch while talker is active

sm0svx committed:
Comment out debug printout

sm0svx committed:
Improved CMake find module for rtl-sdr

sm0svx committed:
Rename functions to avoid name clash

- ProtoVer::major() -> ProtoVer::majorVer()
- ProtoVer::minor() -> ProtoVer::minorVer()

sm0svx committed:
Merge branch 'master' into reflector_tg

sm0svx committed:
Add an Alpine Linux Docker image

sm0svx committed:
Remove unneccessary package from Alpine container image

sm0svx committed:
Fix uninitialized variable

sm0svx committed:
Implement TG selection timeout logic

sm0svx committed:
Print TG# monitoring selection in reflector

sm0svx committed:
Only send MsgTgMonitor if configured

sm0svx committed:
Fix faulty TG selection timeout logic

sm0svx committed:
Update manual pages for the SvxReflector TG feature

sm0svx committed:
Add example for MONITOR_TGS in man page

sm0svx committed:
Fix comment bug

sm0svx committed:
Add code for sending client info (currently disabled)

sm0svx committed:
Add AudioSelector::selectedSource()

New member function AudioSelector::selectedSource() used to find out
which source that are currently selected.

sm0svx committed:
Map CTCSS to TG#

Now possible to map CTCSS to talk groups so that for example a specific talk
group can be selected in the reflector depending on the CTCSS tone
frequency received.

sm0svx committed:
Now possible to use multiple CTCSS squelch tones in parallell

sm0svx committed:
Increase CTCSS_TO_TG detection time to 1s

sm0svx committed:
Make TG selection honor local QSOs

Also reestablish TG selection if the connection is lost to the reflector
server.

sm0svx committed:
Update manual page for CTCSS_TO_TG and CTCSS_FQ

sm0svx committed:
AudioSelector::autoSelectEnabled() const argument

AudioSelector::autoSelectEnabled() can now be called with a constant source.

sm0svx committed:
Stop audio output if no PTT

Audio output will now be stopped until the PTT is actually triggered.

sm0svx committed:
Play announcements through LinkManager

It is now possible to play announcements on all connected logics through
the LinkManager. This is of use when a logic core want to play
announcements on all linked logics.

sm0svx committed:
Add event handler to ReflectorLogic

ReflectorLogic can now play announcements on other connected logics.

sm0svx committed:
Deactivate logic link using prefix only

A logic link is now deactivated using the specified prefix only instead
of using subcommand 0. That make it possible to use command 0 in the
connected logic.

sm0svx committed:
Add a bit of silence before TG report

sm0svx committed:
Also install ReflectorLogic.tcl

sm0svx committed:
Report previous TG when no TG is active

When triggering the TG report using ReflectorLogic subcommand *# and no
TG is currently active, the previously active TG is reported instead.

sm0svx committed:
Quicker response to received TG changes

A sigc signal is now used to signal other logics if the received talk
group updates. This was previously polled once every second so a TG
selection could be sluggish sometimes.

sm0svx committed:
Use larger bw for tone detectors with higher frequency

Automatically select a larger bandwidth (50Hz) for CTCSS_TO_TG tone
detectors that has a detection frequency larger than 300Hz.

sm0svx committed:
Delay TG report until the channel is idle

sm0svx committed:
Always announce TG when entering command

sm0svx committed:
Move most of TG announcement logic to TCL

sm0svx committed:
Suppress excessive TG remote activation announcements

sm0svx committed:
Improved ReflectorLogic event handling

sm0svx committed:
Fix memory hadling bug in ModuleMetarInfo

Memory was freed while still in use. Function onActivity called
ClearWatchMap which erased all FdWatches. That also erased the signal
objects that called the onActivity function. Upon return from onActivity
to the destroyed signal object the module crashed the whole application.
This did not happed on all platforms but typically on Raspberry Pi.

sm0svx committed:
Merge branch 'master' into reflector_tg

sm0svx committed:
Load config variables into TCL for ReflectorLogic

The configuration variables from the ReflectorLogic configuration
section are now loaded into the TCL event handler.

sm0svx committed:
New ReflectorLogic config ANNOUNCE_REMOTE_MIN_INTERVAL

The minimum number of seconds between announcements of the same TG for remote
TG activations. If the same TG is remotely activated repeatedly it will not be
announced until at least the number of seconds specified in this configuration
variable have passed.

sm0svx committed:
Add ReflectorLogic events talker_start and talker_stop

sm0svx committed:
Reflector did not send talker start/stop

The reflector did not send talker start/stop messages for a TG that a
client did not monitor but that the client had currently selected.

sm0svx committed:
Fix ANNOUNCE_REMOTE_MIN_INTERVAL

The ANNOUNCE_REMOTE_MIN_INTERVAL configuration variable did not work so
the defaut time was always used. The default is now 0 so the feature is
disabled if not set in the configuration file.

sm0svx committed:
Only play sound on talker stop for remote talker

In the commented out example code for the talker_stop function the
implementation now only play the sound on remote talker stop, not on
local talker stop.

sm0svx committed:
Indicate talk group selection timeout using a sound

sm0svx committed:
Better talk group timeout sound

sm0svx committed:
Improved multi frequency CTCSS squelch detector

sm0svx committed:
Clear active detector in SquelchCtcss on reset

Also add the m_ prefix to instance variables.

sm0svx committed:
Multiple CTCSS frequency detection in RemoteTrx

sm0svx committed:
Update remotetrx version and man-page

sm0svx committed:
Initial revision of reflector TG QSY

sm0svx committed:
Follow last QSY

- Use command *2 to follow the last QSY.
- Now possible to QSY to specified TG using *1<tg>, e.g. *12405 for TG
  #2405.

sm0svx committed:
Add error check for config variable CTCSS_TO_TG

sm0svx committed:
Improve error prone implementation of CTCSS to TG map

Was using float as index into map which can lead to problems when
rounding errors are present. Now using uint16_t instead.

sm0svx committed:
Fix incorrect time announcement if minute less than 10

If the minute was less than ten, that is the two digit number started
with a zero, a warning was printed about not finding the oX sound clip
and so only the last digit were announced.

sm0svx committed:
Initial revision of Async::HttpServerConnection

sm0svx committed:
Implement talk group priority

Now possible to set priority on monitored talk groups so that a switch
is made to the higher priority talk group if there are activity on it
while being a passive listener on a lower priority talk group.

sm0svx committed:
Add fifo on reflector audio reception

sm0svx committed:
Remove "experimental" labeling from preemphasis

The preemphasis feature was marked as experimental in the documentation
even though it has been in use for years.

sm0svx committed:
Docker Alpine: Add support for GIT_SSL_NO_VERIFY

sm0svx committed:
Prepare release 19.09

sm0svx committed:
Update version for release 19.09

sm0svx committed:
Merge branch 'master' into reflector_tg

sm0svx committed:
Fix TG announcement delay when using multiple logics

This commit revert the changes done to PttCtrl.cpp previously in this
branch.

sm0svx committed:
Don't try to process empty TG selection event

sm0svx committed:
Rework version handling infrastructure for CMake

sm0svx committed:
Merge branch 'master' into reflector_tg

sm0svx committed:
Now possible to use playDtmf in ReflectorLogic

sm0svx committed:
Don't overwrite systemd config files on install

sm0svx committed:
Frn module one way audio bugfix

Closes #444

sm0svx committed:
ASYNC_AUDIO_ALSA_ZEROFILL=1 is not the default

sm0svx committed:
Comma separated values for ReflectorLogic config

Space separated value lists for config of ReflectorLogic now use
comma separated values.

sm0svx committed:
Merge branch 'master' into reflector_tg

sm0svx committed:
Update ChangeLog and version

sm0svx committed:
Improved implementation for SvxLink::SepPair

Improved implementation for SvxLink::SepPair to work better with
streaming operators.

sm0svx committed:
Better compatibility with older g++

sm0svx committed:
Update documentation and systemd config for zerofill change

ASYNC_ALSA_ZEROFILL now default to 1 (enabled) which is now reflected by
the description in the manual page as well as the setting in the systemd
default config file.

sm0svx committed:
Rework ReflectorLogic command structure

The command structure for the ReflectorLogic has now been reworked for
better usability. Each command is prefixed with the prefix specified in
the logic linking configuration but that has been left out in the list
below, as well as the # command termination symbol.

*    -- info/status
1    -- Select previous TG
1nnn -- Select TG nnn
2    -- QSY to "private" TG
2nnn -- QSY to TG nnn
3    -- Follow last QSY

sm0svx committed:
Update documentation and systemd config for zerofill change

ASYNC_ALSA_ZEROFILL now default to 1 (enabled) which is now reflected by
the description in the manual page as well as the setting in the systemd
default config file.

sm0svx committed:
Documentation and default config touch up

sm0svx committed:
Bump version for svxreflector

sm0svx committed:
Disable debug printouts in reflector server

sm0svx committed:
First version of http status for reflector

sm0svx committed:
Add protocol version to status message

sm0svx committed:
Add simple script for monitoring svxreflector status

A simple shell script to monitor svxreflector status in the shell.

watch -cn 0.1 ./svxreflector_status.sh

sm0svx committed:
Add protocol version to status printout

sm0svx committed:
Set current talk group for V1 clients

sm0svx committed:
Fix doc error

sm0svx committed:
Now possible to set the HTTP server port

It is now possible to set the HTTP server port using the HTTP_SRV_PORT
configuration variable.

sm0svx committed:
Bump reflecto version number

sm0svx committed:
Quicker ReflectorLogic reconnect

sm0svx committed:
Better random reconnect time

sm0svx committed:
Better HTTP version handling

sm0svx committed:
Better naming for initial request HTTP state

STATE_EXPECT_METHOD -> STATE_EXPECT_START_LINE

sm0svx committed:
HTTP server code documentation cleanup

sm0svx committed:
Use JSON for Voter:sql_state

sm0svx committed:
Client info initial revision

Add support for static client info and transmission of signal strength
values for receivers.

sm0svx committed:
Client info fully functional

sm0svx committed:
Fixes for building on older distros

sm0svx committed:
Fix misspelled JSONCPP_INCLUDE_DIRS for CMake

sm0svx committed:
Add the client_info.json file and some docs

sm0svx committed:
New state event LocalRx:sql_state

sm0svx committed:
Rename "client info" to "node info"

sm0svx committed:
Send signal strenght updates through TCP

Switch to sending signal strength updates through TCP instead of UDP so
that we don't risk loosing a sql close.

sm0svx committed:
Fix warning for MsgSignalLevelValuesBase

sm0svx committed:
Set rx active with sql open for normal receivers

sm0svx committed:
State event for TX status

Information about TX status, e.g. if transmitting or not, is now
published as a state event and is also sent to the reflector server.

sm0svx committed:
Remove debug printout for MsgTxStatus

sm0svx committed:
The reflector server can now handle newer clients

The reflector server now have a mechanism to handle clients using a newer
protocol version than the server supports. If a client upon connect
indicates a protocol version newer than what the server can handle it
will ask the client to downgrade its protocol version. It is up to the
client to choose wether it can handle the downgrade or not.

sm0svx committed:
Mute first reflector TX after TG activation

By default mute the first transmission after selecting a talk group.
This is good for reducing unneccessary talk group activations and to
allow local node DTMF command submission without disturbing an ongoing
QSO on a talk group.

sm0svx committed:
Add code doc for LogicBase::setReceivedTg

sm0svx committed:
Move misplaced node_info.json

The default node_info.json was by mistake placed in the reflector
directory instead of the svxlink directory.

sm0svx committed:
Add a nodeLocation field in node_info.json

sm0svx committed:
Add TX_ID doc in man-page

sm0svx committed:
Add "hidden" property to node_info.json

sm0svx committed:
Rx:sql_state now compatible with remote receivers

The Rx:sql_state was renamed from LocalRx:sql_state to be more general
and is now also emitted from a NetRx.

sm0svx committed:
Now possible to set TX_ID on a NetTx

This really is a workaround so that we don't have to change the
RemoteTrx protocol at this moment. Instead of getting the TX_ID from the
remote transmitter we set it in the NetTx configuration section.

sm0svx committed:
Never set previous TG to 0

sm0svx committed:
Split MUTE_FIRST_TX in local and remote versions

The MUTE_FIRST_TX ReflectorLogic configuration variable has been split
into two new variables, MUTE_FIRST_TX_LOC and MUTE_FIRST_TX_REM. This
has been done since the muting on local talk group activation is quite
intuitive but muting due to remote talk group activivation is not.

Have a look in the svxlink.conf (5) manual page for more information.

sm0svx committed:
Better announcements for QSY

- New soundclips needed: Core/qsy, Core/ignored

sm0svx committed:
Add DTMF controlled temporary monitoring

It is now possible to use DTMF command <prefix>4<tg> to temporarily
monitor a talk group on a node. The monitored talk group will time out
after one hour.

sm0svx committed:
Add info printouts for temporary TG monitoring events

sm0svx committed:
Settable TMP_MONITOR_TIMEOUT

sm0svx committed:
Merge branch 'reflector_tg'

The reflector_tg branch add support for switching traffic on talk groups
in the SvxReflector.

sm0svx committed:
Update travis CI specification

- Use Ubuntu Bionic
- Add package libjsoncpp-dev

sm0svx committed:
Remove node IP address from HTTP status data

sm0svx committed:
New signal level "detector": CONST

sm0svx committed:
Fix bugs in SigLevDetConst

- RX_ID was not read
- The siglev update timer was fixed at one second

dl1hrc committed:
merged with master

sm0svx committed:
ModuleEchoLink LINK_IDLE_TIMEOUT bugfix

ModuleEchoLink bugfix: The LINK_IDLE_TIMEOUT did not work if the
remote_timeout TCL function was changed to not play any sound.

sm0svx committed:
Improve CPack for DEB packages

- Properly mark configuration files
- Set owner:group for non-root files
- Create svxlink user:group

sm0svx committed:
Silence deb postinstall script

sm0svx committed:
ReflectorLogic bugfix: Assertion if no LinkManager

SvxLink would crash on an assertion if a ReflectorLogic was defined but
no logic links were defined.

sm0svx committed:
ModuleEchoLink LINK_IDLE_TIMEOUT bugfix

ModuleEchoLink bugfix: The LINK_IDLE_TIMEOUT did not work with short
timeout values where the timeout was shorter than the length of a
locally received transmission.

sm0svx committed:
Flatten install path for html man pages

sm0svx committed:
Add better SvxReflector status script

dl1hrc committed:
Bugfix for sending metar info twice although requesting it one time

dl1hrc committed:
Merge remote-tracking branch 'sm0svx/master' into DmrLogic_experimental

sm0svx committed:
Update version and ChangeLog

sm0svx committed:
Merge branch 'dl1hrc-MetarDoubleRequestBug'

dl1hrc committed:
bugfix: crash on long LocationInfo comments

dl1hrc committed:
Merge remote-tracking branch 'sm0svx/master' into DmrLogic_experimental

sm0svx committed:
Better warning printouts for illegal UDP datagrams

sm0svx committed:
Now possible to silently drop all incoming Echolink connections

It's now possible to silently ignore all incoming connections using the
DROP_ALL_INCOMING configuration variable.

dl1hrc committed:
Merge remote-tracking branch 'sm0svx/master' into DmrLogic_experimental

sm0svx committed:
Subscribe to config changes now possible

The receiver configuration variables SQL_HANGTIME, SQL_EXTENDED_HANGTIME
and SQL_EXTENDED_HANGTIME_THRESH are now settable from within a C++
module for example.

This is made possible by the addition of the Config::valueUpdated signal
that is emitted when the Config::setValue function is called.

dl1hrc committed:
Merge remote-tracking branch 'sm0svx/master' into DmrLogic_experimental

sm0svx committed:
UTF-8 bugfix in svxreflector_status.py

Dloranger committed:
Update ModuleTxFan.tcl

Updating discussion to match the actual module and provide guidance on how to configure the module

dl1hrc committed:
ModuleMetarInfo: more flexible announcement of an airport at the beginning of a metar report

dl1hrc committed:
fixed some spaces

dl1hrc committed:
fixed some spaces

dl1hrc committed:
Merge remote-tracking branch 'sm0svx/master' into DmrLogic_experimental

sm0svx committed:
Update cpp/h template files

sm0svx committed:
Improved GPIO up/down scripts

Improved GPIO up/down scripts with more error handling and support for
handling more cases.

sm0svx committed:
Add chunked mode to http connections

sm0svx committed:
Fix small memory deallocation bug in Async::AudioEncoderOpus

sm0svx committed:
New template class Async::Factory for creating object factories

sm0svx committed:
Improve usability for the Async::Factory class

- Introduce variadic templates to support classes with constructors that
  take one or more arguments.
- Better class naming
- Better documentation
- Add a detaild example also showing usage of constructor arguments.

sm0svx committed:
Add an AudioContainer framework

sm0svx committed:
The svxreflector_status.py script now use curses

The curses library is now used to make the svxreflector_status.py script
a fullscreen application that update the view a couple of times per
second.

sm0svx committed:
Use lambda function instead of anonymous namespace function

sm0svx committed:
Move svxreflector_status.py to the reflector directory

The svxreflector_status.py script have now been moved from the scripts
directory to the reflector directory. It is also installed as part of a
"make install".

sm0svx committed:
New SvxReflector feature: Auto QSY

It is now possible to set a per talkgroup time in the reflector after
which a QSY to a random talkgroup is performed. This can typically be
used on call channels and wide area channels so that stations do not
dwell there for too long.

sm0svx committed:
Don't try to build ogg-stuff if dependencies are not available

sm0svx committed:
Fix compiler warning

sm0svx committed:
An Auto QSY is now only triggered after talker stop

sm0svx committed:
Update copyright notices

sm0svx committed:
Bugfix: Some squelch types did not report the integrated siglev

Squelch types DDR, SIM and TONE reported the "raw" signal strength
instead of the integrated one. This fix can change the behavior of the
voter and the extended hangtime feature.

sm0svx committed:
Add logical negation to the COMBINE squelch

It now is possible to write expressions like the one below for
SQL_DET=COMBINE.

  SQL_COMBINE=(Rx1:SIGLEV | Rx1:CTCSS) & !Rx1:CTCSS_BLOCK

sm0svx committed:
Check channel number when opening audio device

The configured channel number for an audio device is now checked against
the value set in CARD_CHANNELS before trying to open the audio device.

sm0svx committed:
Remove debug printout

sm0svx committed:
Add debug mode for the CTCSS detector

This commit add a debug mode for the CTCSS detector. Enable by setting
CTCSS_DEBUG to 1 in a receiver configuration section.

sm0svx committed:
Fix sign bug in CTCSS_DEBUG

sm0svx committed:
Use bash instead of sh in filter_sounds

The same with play_sound.

sm0svx committed:
Fix ReflectorLogic event buf with tg_qsy_failed

The event tg_qsy_failed was called tg_qsy_idle in the C++ source code
while it was called tg_qsy_failed in TCL.

sm0svx committed:
Update templates

dermodmaster committed:
Update eventsource.pl

Fix parse method for non json values

dermodmaster committed:
Update eventsource.pl

sm0svx committed:
ReflectorLogic UDP heartbeat interval adjustment

It is now possible to adjust the ReflectorLogic UDP heartbeat interval
using the UDP_HEARTBEAT_INTERVAL configuration variable.

sm0svx committed:
Config::valueUpdated only emitted if value changed

sm0svx committed:
A Pty can now be set in line buffered mode

The Pty::setLineBuffered method can now be used to enable line buffered mode
where incoming chars are buffered until a <CR> or <LF> is received.

sm0svx committed:
Extended support for changing variables at runtime

- Now possible to set configuration variables from TCL using the
  setConfigValue function. This can be used to change certain configuration of
  SvxLink while it is running. Only some configuration variables support being
  changed at runtime.
- If a configuration variable is changed in the logic core, or in the main
  configuration section for a module, the new TCL function config_updated will
  be called.
- New configuration variable ONLINE that can be used to decide if a logic core
  should be active or not on program startup. This variable also support being
  set at runtime.
- Use the new configuration variable COMMAND_PTY to set up a PTY that can be
  used to send commands to a logic core.

sm0svx committed:
Fix JSON data getting lost

The JSON data never reached the JSON parser since it was "eaten" by the
assignment to $rx_info_raw.

sm0svx committed:
Fix eventsource.pl to handle both JSON and simple formats

Merge pull request #519 from dermodmaster/patch-1. Fix parse method for non JSON values in eventsource.pl. The script would fail if non JSON data was received.

sm0svx committed:
Eventsource script improvements

Several improvements were made to the eventsource.pl script and the web
pages.

- Now support the Rx:sql_state and Tx:state events
- The Logic:transmit event no longer have to be added
- Paths can now be set on the command line for eventsource.pl
- Add a simple web server script that can be used to try out the system
  locally
- Add missing icons
- Add jQuery
- More generic and trimmed down setup in the JSON files

dl1hrc committed:
trying to merge with upstream

dl1hrc committed:
trying to merge with upstream

dl1hrc committed:
merged with upstream (it was really a lot of work :/)

dl1hrc committed:
merged with upstream (it was really a lot of work :/)

dl1hrc committed:
some smaller changes for merge

sm0svx committed:
Filter out CTCSS band for voice output

Interference on CTCSS transmission could be caused by low frequency
content in audio clips or incoming audio on Echolink for example.

sm0svx committed:
Ignore prioritized talkgroups qfter QSY

Ignore prioritized talkgroups qfter QSY if there have been local node
activity prior to the QSY. This fix a problem where a node would switch
to a prioritized talkgroup after a QSY to a lower priority talkgroup, if
there have been no local activity after the QSY. Priority will be
honored again after the node have switched to TG#0.

sm0svx committed:
Only clear the local activity flag if idle or TG#0

sm0svx committed:
Configurable audio limiter for both RX and TX audio

- Make the audio limiter threshold in the receiver configurable using the
  LIMITER_THRESH configuration variable in a receiver configuration section.
- Add an audio limiter to the transmitted audio path. The threshold may be set
  using the LIMITER_THRESH configuration variable in a TX configuration
  section.

dl1hrc committed:
Merge remote-tracking branch 'sm0svx/master' into DmrLogic_experimental
